services/web/server/src/simcore_service_webserver/director_v2_core_dynamic_services.py

```diff
@@ -0,0 +1,218 @@
+    settings: DirectorV2Settings = get_plugin_settings(app)
+    backend_url = settings.base_url / f"dynamic_services/{node_uuid}"
+
+    service_state = await request_director_v2(
+        app, "GET", backend_url, expected_status=web.HTTPOk
+    )
+
+    assert isinstance(service_state, dict)  # nosec
+    return service_state
+
+
+@log_decorator(logger=log)
+async def run_dynamic_service(
+    app: web.Application,
+    user_id: PositiveInt,
+    project_id: str,
+    service_key: str,
+    service_version: str,
+    service_uuid: str,
+    request_dns: str,
+    request_scheme: str,
+    service_resources: ServiceResourcesDict,
+) -> DataType:
+    """
+    Requests to run (i.e. create and start) a dynamic service:
+    - legacy services request is redirected to `director-v0`
+    - dynamic-sidecar `director-v2` will handle the request
+    """
+    data = {
+        "user_id": user_id,
+        "project_id": project_id,
+        "key": service_key,
+        "version": service_version,
+        "node_uuid": service_uuid,
+        "basepath": f"/x/{service_uuid}",
+        "service_resources": ServiceResourcesDictHelpers.create_jsonable(
+            service_resources
+        ),
+    }
+
+    headers = {
+        "X-Dynamic-Sidecar-Request-DNS": request_dns,
+        "X-Dynamic-Sidecar-Request-Scheme": request_scheme,
+    }
+
+    settings: DirectorV2Settings = get_plugin_settings(app)
+    started_service = await request_director_v2(
+        app,
+        "POST",
+        url=settings.base_url / "dynamic_services",
+        data=data,
+        headers=headers,
+        expected_status=web.HTTPCreated,
+    )
+
+    assert isinstance(started_service, dict)  # nosec
+    return started_service
+
+
+@log_decorator(logger=log)
+async def stop_dynamic_service(
+    app: web.Application, service_uuid: str, save_state: bool = True
+) -> None:
+    # stopping a service can take a lot of time
+    # bumping the stop command timeout to 1 hour
+    # this will allow to sava bigger datasets from the services
+    settings: DirectorV2Settings = get_plugin_settings(app)
+    await request_director_v2(
+        app,
+        "DELETE",
+        url=(settings.base_url / f"dynamic_services/{service_uuid}").update_query(
+            save_state="true" if save_state else "false",
+        ),
+        expected_status=web.HTTPNoContent,
+        timeout=settings.DIRECTOR_V2_STOP_SERVICE_TIMEOUT,
+    )
+
+
+@log_decorator(logger=log)
+async def stop_dynamic_services_in_project(
+    app: web.Application,
+    user_id: Optional[PositiveInt] = None,
+    project_id: Optional[str] = None,
+    save_state: bool = True,
+) -> None:
+    """Stops all services of either project_id or user_id in concurrently"""
+    running_dynamic_services = await get_dynamic_services(
+        app, user_id=user_id, project_id=project_id
+    )
+
+    services_to_stop = [
+        stop_dynamic_service(
```

Comment:
here, i see something in the long term needing enhancements (maybe add my ramblings in condensed form as a FIXME/TODO...):
- What to we expect to happen when a study is closed and his call is started, but something goes wrong? I guess for now, the individual stop_dynamic_services calls might fail and raise, and there will be no retry?
-  What if the user has 100 services, and one of the stop_dynamic_service calls to the director-v2 times out? Do we handle this gracefully?
- Scenario: A dynamic service is not reachable. Do we assume it is dead? Then, we fire-and-forget and ignore that the director_v2 will not be able to handle the stop_dynamic_service request, as the service is already out-ot-reach. Alternatively, we might assume the service only to be unreachable temporarily (load on the machine, network issues, unluck, ...). Then, we might want to retry.

I guess what I am trying to say in a nutshell is that in my mind, either the director-v2 is really resilient in handling these requests, even if weird failures occur, and the director-v2 is always reachable and responsive, or I would expect this codeblock here to have a bit more logic and handling to provide robustness.

Open for discussion I guess, not actionable for now.

Reply:
pair reviewed